### PR TITLE
add filter

### PR DESCRIPTION
**Pull request by SergeyOdnosteblets**

### CODE REVIEW

File: src/components/Filter/Filter.module.scss.d.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const form: string;
+export const form__input: string;
+export const form__error: string;
+export const form__button:string
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.module.scss.d.ts]: formatting

File: src/components/Filter/Filter.tsx [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import React from 'react';
+import { useForm } from 'react-hook-form';
+import { FilterTypes } from '../../types/FilterTypes';
+import { FilterTypesForm } from '../../types/FilterTypesForm';
+import { UserInfo } from '../../types/UserInfo';
+
+import styles from './Filter.module.scss';
+
+export const Filter: React.FC<FilterTypes> = ({ list, setIsFilter }) => {
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: what does `setIsFilter ` mean?

File: src/components/Filter/Filter.tsx [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    if (data.gender) {
+      let filter = list.filter((item: UserInfo) => {
+        return item.gender === data.gender;
+      });
+      setIsFilter(filter);
+    }
+
+    if (data.ageFrom && data.ageTo) {
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: What if I want to get people where age is more than 5?
It won't work

File: src/components/Filter/Filter.tsx [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    formState: { errors },
+    handleSubmit,
+    reset,
+  } = useForm<FilterTypesForm>({
+    mode: 'onSubmit',
+  });
+
+  const onSubmit = (data: FilterTypesForm) => {
+    if (data.firstName) {
+      let filter = list.filter((item: UserInfo) => {
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: that's  not filter
filteredList

File: src/components/Filter/Filter.tsx [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+      });
+      setIsFilter(filter);
+    }
+
+    if (data.ageFrom && data.ageTo) {
+      let filter = list.filter((item: UserInfo) => {
+        return item.age > data.ageFrom && item.age < data.ageTo;
+      });
+      setIsFilter(filter);
+    }
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: we should be able to filter the list with different options
eg age > 5 && gender = male && firstName includes "b"

File: src/components/ListItem/ListItem.tsx [hunk 6]
```diff
@@ -20,26 +25,35 @@ export const ListItem: React.FC<UserListItemProps> = ({ list, removeUser, handle
...
-  return (
-    list &&
-    list.map((item: UserInfo) => {
-      return (
-        <div className={styles.main} key={item.id}>
-          <div className={styles.item}>
-            <div className={styles.user} onClick={() => handleClick(item)} data-testid="user-id">
-              {item.firstName}
+  return isFilter.length
+    ? isFilter.map((item: UserInfo) => {
```
Comment (t-shevchuk) [inline src/components/ListItem/ListItem.tsx]: I don't like that

File: src/components/UserModal/UserModal.tsx [hunk 7]
```diff
@@ -6,8 +6,8 @@ import { UserInfo } from '../../types/UserInfo';
 import styles from './UserModal.module.scss';
 
 export const UserModal: React.FC<UserEdit> = ({
-  setIsModalActive,
-  isModalActive,
+  setIsModal,
```
Comment (t-shevchuk) [inline src/components/UserModal/UserModal.tsx]: That's always modal

**Review by t-shevchuk — CHANGES_REQUESTED** [review]

File: src/components/Filter/Filter.tsx (hunk 2 above)
Comment (SergeyOdnosteblets) [reply src/components/Filter/Filter.tsx]: changed it

File: src/components/Filter/Filter.tsx (hunk 3 above)
Comment (SergeyOdnosteblets) [reply src/components/Filter/Filter.tsx]: changed it

File: src/components/ListItem/ListItem.tsx [hunk 8]
```diff
@@ -19,27 +24,44 @@ export const ListItem: React.FC<UserListItemProps> = ({ list, removeUser, handle
...
-    list &&
-    list.map((item: UserInfo) => {
-      return (
-        <div className={styles.main} key={item.id}>
-          <div className={styles.item}>
-            <div className={styles.user} onClick={() => handleClick(item)} data-testid="user-id">
-              {item.firstName}
+  
+  return filterUsers.length
+    ? filterUsers.map((item: UserInfo) => {
```
Comment (t-shevchuk) [inline src/components/ListItem/ListItem.tsx]: 1. firlteredUsers
2. don't use ternary operator here

Comment (t-shevchuk) [reply src/components/ListItem/ListItem.tsx]: duplicated code again

File: src/pages/List/List.tsx [hunk 9]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
     setIsModalActive(!isModalActive);
   };
 
+  console.log(filterUsers);
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: remove

File: src/pages/List/List.tsx [hunk 10]
```diff
@@ -1,96 +1,16 @@
...
-      id: '636cb4c0-7390-41e2-8eee-50ae612987c5',
-      firstName: 'Rob',
-      lastName: 'Smith',
-      age: 34,
-      gender: 'Male',
-      country: 'United States',
-    },
-  ]);
-
+export const List: React.FC<ListTypes> = ({ list, setList }) => {
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: Listcontainer

File: src/pages/List/List.tsx [hunk 11]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
     setIsModalActive(!isModalActive);
   };
 
+  console.log(filterUsers);
+
   return (
     <div className={styles.main} data-testid="all-user-id">
-      <button className={styles.button} onClick={() => handleEdit()}>
-        Add Users
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx:114]: user

File: src/pages/List/List.tsx [hunk 12]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
...
-        <UserModal
-          setIsModalActive={setIsModalActive}
-          isModalActive={isModalActive}
+      <div className={styles.filter}></div>
+      <Filter setFilterUsers={setFilterUsers} list={list} />
+      <div className={styles.list}>
+        <button className={styles.button} onClick={() => handleEdit()}>
+          Add Users
+        </button>
+        <ListItem
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: List

File: src/pages/List/List.tsx [hunk 13]
```diff
@@ -1,96 +1,16 @@
...
-      age: 34,
-      gender: 'Male',
-      country: 'United States',
-    },
-  ]);
-
+export const List: React.FC<ListTypes> = ({ list, setList }) => {
   const [userToEdit, setUserToEdit] = useState<UserInfo | null>(null);
   const [isModalActive, setIsModalActive] = useState(false);
+  const [filterUsers, setFilterUsers] = useState<UserInfo[]>([]);
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: filteredUsers = useState(list)

File: src/pages/List/List.tsx [hunk 14]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
...
-      <button className={styles.button} onClick={() => handleEdit()}>
-        Add Users
-      </button>
-      <ListItem list={list} removeUser={removeUser} handleEdit={handleEdit} />
-      {isModalActive && (
-        <UserModal
-          setIsModalActive={setIsModalActive}
-          isModalActive={isModalActive}
+      <div className={styles.filter}></div>
+      <Filter setFilterUsers={setFilterUsers} list={list} />
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: pass function
(data) => {
 filter here 
setFilteredUsers
}

File: src/pages/List/List.tsx [hunk 15]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
...
+      <div className={styles.list}>
+        <button className={styles.button} onClick={() => handleEdit()}>
+          Add Users
+        </button>
+        <ListItem
           list={list}
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: list={filteredUsers}

File: src/pages/List/List.tsx [hunk 16]
```diff
@@ -108,21 +28,32 @@ export const List: React.FC = () => {
...
-          setList={setList}
-          userToEdit={userToEdit}
+          removeUser={removeUser}
+          handleEdit={handleEdit}
+          filterUsers={filterUsers}
```
Comment (t-shevchuk) [inline src/pages/List/List.tsx]: remove

File: src/components/Filter/Filter.tsx [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+import { FilterTypes } from '../../types/FilterTypes';
+import { UserInfo } from '../../types/UserInfo';
+
+export const Filter: React.FC<FilterTypes> = ({ list, setFilterUsers }) => {
+  const { register, handleSubmit, reset } = useForm<FilterFormTypes>({
+    mode: 'onBlur',
+    defaultValues: {},
+  });
+
+  const onSubmit = (data: FilterFormTypes) => {
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: will be moved

Comment (t-shevchuk) [reply src/components/Filter/Filter.tsx]: list = list ?. filter ((users) => !data.firstName || item.firstName.toLowerCase().includes(data.firstName))
?.filter(... age from )
?.filter(... age to  )
?.filter(... gender) ?? []

File: src/components/Filter/Filter.tsx [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+            className={styles.form__input}
+            type="number"
+            {...register('ageTo', {
+              min: 0,
+              max: 100,
+            })}
+            placeholder="Age To"
+          />
+
+          <select className={styles.form__input} {...register('gender')}>
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: multi-select here

File: src/components/Filter/Filter.tsx [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+            <option value="">gender</option>
```
Comment (t-shevchuk) [inline src/components/Filter/Filter.tsx]: not option

File: src/types/UserModalTypes.ts [hunk 20]
```diff
@@ -2,7 +2,7 @@ import { UserInfo } from "./UserInfo";
 
 export interface UserModalTypes {
     isModalActive: boolean;
-    setIsModalActive: React.Dispatch<React.SetStateAction<boolean>>;
+    setIsModal: React.Dispatch<React.SetStateAction<boolean>>;
```
Comment (t-shevchuk) [inline src/types/UserModalTypes.ts]: naming

File: src/types/UserToEdit.ts [hunk 21]
```diff
@@ -1,10 +1,10 @@
...
 
-export interface UserEdit {
+export interface UserToEdit {
     userToEdit?: UserInfo | null
     setIsModalActive: React.Dispatch<React.SetStateAction<boolean>>
     isModalActive: boolean
     list: UserInfo[];
     setList: (list: UserInfo[]) => void 
-    setUserToEdit?: (list: UserInfo | null) => void 
+    setEditUser?: (list: UserInfo | null) => void
```
Comment (t-shevchuk) [inline src/types/UserToEdit.ts]: naming

**Review by t-shevchuk — CHANGES_REQUESTED** [review]

File: src/pages/ListContainer/ListContainer.tsx [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+  const removeUser = (userObj: UserInfo) => {
+    let newList = [...list];
+    newList.splice(list.indexOf(userObj), 1);
+
+    setList(newList);
+  };
+
+  const handleEdit = (userObj?: UserInfo) => {
+    if (userObj) {
```
Comment (t-shevchuk) [inline src/pages/ListContainer/ListContainer.tsx]: You don't need to double-check userObject

    if (userObj) {
     setUserToEdit(userObj);
    } else {
      setUserToEdit(null);
    }

File: src/types/FilterFormTypes.ts [hunk 23]
```diff
@@ -0,0 +1,8 @@
+export interface FilterFormTypes {
+    firstName?: string;
+    ageFrom?: number;
+    ageTo?: number;
+    gender?: string;
+    test?: string[]
```
Comment (t-shevchuk) [inline src/types/FilterFormTypes.ts]: What is the test?

File: src/types/FilterTypes.ts [hunk 24]
```diff
@@ -0,0 +1,8 @@
+import { FilterFormTypes } from "./FilterFormTypes"
+import { UserInfo } from "./UserInfo"
+
+export interface FilterTypes {
+    setFilteredUsers: React.Dispatch<React.SetStateAction<UserInfo[]>>
+    onSubmit: (data: FilterFormTypes) => void
+
```
Comment (t-shevchuk) [inline src/types/FilterTypes.ts]: fix formatting

File: src/types/UserListItemProps.ts [hunk 25]
```diff
@@ -1,8 +1,8 @@
 import { UserInfo } from "./UserInfo";
 
 export interface UserListItemProps {
-    list: any
     removeUser: (userObj: UserInfo) => void 
     handleEdit: (userObj: UserInfo) => void
+    firlteredUsers: any
```
Comment (t-shevchuk) [inline src/types/UserListItemProps.ts]: UserInfo[]

File: src/types/UserToEdit.ts [hunk 26]
```diff
@@ -1,10 +1,10 @@
...
 
-export interface UserEdit {
+export interface UserToEdit {
     userToEdit?: UserInfo | null
     setIsModalActive: React.Dispatch<React.SetStateAction<boolean>>
     isModalActive: boolean
     list: UserInfo[];
     setList: (list: UserInfo[]) => void 
-    setUserToEdit?: (list: UserInfo | null) => void 
+    setToEditUser?: (list: UserInfo | null) => void
```
Comment (t-shevchuk) [inline src/types/UserToEdit.ts]: setUserToEdit 
English grammatical

**Review by t-shevchuk — CHANGES_REQUESTED** [review]

File: src/pages/ListContainer/ListContainer.tsx (hunk 22 above)
Comment (SergeyOdnosteblets) [reply src/pages/ListContainer/ListContainer.tsx:24]: changed

File: src/types/FilterFormTypes.ts (hunk 23 above)
Comment (SergeyOdnosteblets) [reply src/types/FilterFormTypes.ts]: deleted it

**Review by t-shevchuk — APPROVED** [review]

**Pull request merged**